modules/python/python-http-header.c

```diff
@@ -0,0 +1,383 @@
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  PyObject *py_list = (PyObject *) user_data;
+  if (PyList_Append(py_list, py_str) != 0)
+    {
+      gchar buf[256];
+
+      msg_error("Error adding new item to Python List",
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+    }
+
+exit:
+  Py_XDECREF(py_str); // Even if append() succeeds, it doesn't steal the reference to py_str.
+}
+
+static PyObject *
+_py_convert_list_to_pylist(List *list)
+{
+  PyObject *py_list = PyList_New(0);
+  g_assert(py_list);
+
+  if (list)
+    list_foreach(list, _py_append_str_to_pylist, py_list);
+
+  return py_list;
+}
+
+static gboolean
+_py_attach_bindings(PythonHttpHeaderPlugin *self)
+{
+  PyObject *py_args = NULL;
+
+  self->py.class = _py_resolve_qualified_name(self->class);
+  if (!self->py.class)
+    {
+      gchar buf[256];
+
+      msg_error("Error looking up Python class",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  py_args = _py_create_arg_dict(self->options);
+  if (!py_args)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating argument dictionary",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.instance = _py_invoke_function(self->py.class, py_args, self->class, self->super.name);
+  if (!self->py.instance)
+    {
+      gchar buf[256];
+
+      msg_error("Error instantiating Python class",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.get_headers = _py_get_attr_or_null(self->py.instance, "get_headers");
+  if (!self->py.get_headers)
+    {
+      msg_error("Error initializing plugin, required method not found",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"));
+    }
+
+exit:
+  Py_XDECREF(py_args);
+  return self->py.get_headers != NULL;
+}
+
+static void
+_py_detach_bindings(PythonHttpHeaderPlugin *self)
+{
+  Py_CLEAR(self->py.class);
+  Py_CLEAR(self->py.instance);
+  Py_CLEAR(self->py.get_headers);
+}
+
+static void
+_append_str_to_list(gpointer data, gpointer user_data)
+{
+  List *list = (List *) user_data;
+  list_append(list, data);
+}
+
+static void
+_append_headers(PythonHttpHeaderPlugin *self, HttpHeaderRequestSignalData *data)
+{
+  PyObject *py_ret = NULL,
+            *py_list = NULL,
+             *py_args = NULL,
+              *py_ret_list = NULL;
+
+  time_t now = time(NULL);
+  if ((now - self->last_run) < self->timeout)
```

Comment:
Do we need this timeout functionality in the framework? This is something that could be implemented in the user code. In a complex scenario, when some headers can be cached while others not, users cannot give universal timeout. Performance wise, I do not think a python side caching would measure significantly either.

My main problem is that I feel unconfortable assigning a timeout parameter to a `response`, because it typically means something else. This of course could be resolved by renaming too (like cache-time or similar), but first I am interested if we want to have it at all.

Reply:
@furiel: this is not a python-side caching... and thus it could have a real performance benefit.
The naming is not the best, and yes, there are better alternatives. But... when the one and only reason for existence of your plugin is to implement a http auth header that has an expiration time measured in hours then this caching mechanism implemented in the C side is a good thing.

Reply:
I would also prefer performance, but I am still not convinced we could measure performance difference. A http request is done in the milliseconds range at best, due to network latency. A call to a python function is in microseconds range.

If we could do the caching in python, then we could invent very neat APIs. Also, users could use nice caching libraries (I randomly found https://pypi.org/project/expiringdict/), or we could provide our own.

For example users could write a non-caching
```
headers = {"hdr": "value"}
return headers
```
and with caching with our domain specific decorator
```
@cache_timeout(3600)
headers = {"hdr" : "value"}
return headers
```

Less c and more python would scale better if we want to add another feature in the future. For example we might want to add a cache cancellation function too. For example if the credentials file is changed on the disk, users might want to drop cache. Or in case 403 error code has arrived as a response to the http destination.

Reply:
We touched this topic yesterday IRL but don’t know this comment is added after or before our conversation...

Yes, we could implement more and more features on top of the base concept. But... you perfectly know that this PR opened with reason. So this is not a l’art pour l’art PR where I want to create the Holy Python Http Header module, with decorators and others... 

Shortly: anyone can easily add improvements (or features) on top of it, modify in another PR.

On the performance ...

We are heavy working on Python support.
Maybe I’m wrong, but I’m expecting more Python modules added to syslog-ng than before.
So when we are talking about performance (related to this PR), we should talk about concurrency and parallelism  ... in syslog-ng there are elements in the processing pipeline that are able to run in a parallel way. When we are implementing these elements in Python, it’s not true anymore (unless you run each Python module in a separate process and add an IPC mechanism to syslog-ng ...).
Yes, I’m talking about GIL (and yes, i know about IronPython, etc... but that would be unfair to cite those implementations here: we are using CPython).
So I think when we are talking about performance we should think about E2E, real life scenarios: and when you have more  and more python modules, it’s a good pattern if you cache values to avoid holding the GIL - so it’s not only about how to speed up (...) the HTTP requests by skipping the C->Python calls but about trying to maximalize the number of Python modules running in parallel way.

I’m thinking about split the caching part from this PR as it’s just a perf. improvement...

Reply:
Yes I added this comment before our irl discussion. And I agree with you.

In the end, users can still disable c-side caching if they want (by returning zero), and add their own caching implementation python side. So it is fine for me to have this kind of c-side caching.